File: Code/C++/insertion_sort.cpp
/*
Insertion sort is a simple and efficient comparison sort. In this algorithm, each iteration removes
an element from the input data and inserts it into the correct position in the list being sorted. The
choice of the element being removed from the input is random and this process is repeated until
all input elements have gone through.

 * Algorithm :
	Every repetition of insertion sort removes an element from the input data, and inserts it into the
	correct position in the already-sorted list until no input elements remain. Sorting is typically done
	in-place. The resulting array after k iterations has the property where the first k + 1 entries are
	sorted.Each element greater than x is copied to the right as it is compared against x.
*/

#include <iostream>
using namespace std;

int main()
{
	int n;
	cin >> n;
	int a[n];
	for(int i=0;i<n;i++){
       cin >> a[i];
	}
	
	for(int i = 1;i<n ;i++ ){
		for(int j = i; j>0 ; j-- ){
			if(a[j]<a[j-1]){
				swap(a[j],a[j-1]);
			}
			else{
				break;
			}
		}
	}

	for(int i=0;i<n;i++){
		cout<<a[i]<<" ";
	}


	return 0;
}

/* 
Test case : 
Input - 5
4 3 5 1 2

Output : 
1 2 3 4 5

Time Complexity : 
	Worst case complexity: Θ(n^2 )
	Best case complexity: Θ(n)
	Average case complexity: Θ(n^2 )
	Worst case space complexity: O(n^2 ) total, O(1) auxiliary
*/

File: Code/Java/Insertion_sort.java
/*
Insertion sort is a sorting algorithm.
The array is virtually divided into two sections i.e., a sorted and an unsorted section. 
Values from the unsorted section are picked and placed at the correct 
position in the sorted section.

*/

import java.util.*;
public class Insertion_sort
{
    public static void main(String args[])
    {
        int k = 0;
        int n=0;
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the array size ");
        n = sc.nextInt();
        int A[] = new int [n];
        System.out.println("Enter the array elements ");
        for(k=0;k<n;k++)
        {
            A[k]=sc.nextInt();
        }
        for (int a = 1; a < n; a++) 
        {  
            int c = A[a];  
            int b = a-1;  
            while ( (b > -1) && ( A[b] > c ) ) 
            {  
                A[b+1] = A[b];  
                b--;  
            }  
            A[b+1] = c;  
        }
        System.out.println("Sorted array: ");
        for(k=0; k<n; k++)
        {
            System.out.println(A[k]);
        }
    }
}

/*

TEST CASE 1:
INPUT

Enter the array size 
5
Enter the array elements 
25
46
38
96
1

OUTPUT:

Sorted array: 
1
25
38
46
96

TEST CASE 2:
INPUT:

Enter the array size 
8
Enter the array elements 
2
56
1
36
99
1000
85
1 

OUTPUT:

Sorted array: 
1
1
2
36
56
85
99
1000

Time complexity: 

Worst-case: О(n^2) comparisons and swaps
Best-case: O(n) comparisons and O(1) swaps
Average-case: O(n2) comparisons and swaps

Worst-case space complexity: О(n) total, O(1) auxiliary

*/